File: Divide-and-Conquer Searching/Binary Search/Last Occurance Search/LastOccurenceSearch.c
#include <stdio.h>

int lastOccurrenceSearch(int arr[], int size, int target)
{
    int left = 0;
    int right = size - 1;
    int answer = -1;

    while (left <= right)
    {
        int mid = left + (right - left) / 2;

        if (arr[mid] == target)
        {
            answer = mid;
            left = mid + 1;
        }
        else if (arr[mid] < target)
        {
            left = mid + 1;
        }
        else
        {
            right = mid - 1;
        }
    }

    return answer;
}

int main()
{
    int arr[] = {1, 3, 3, 3, 5, 7, 9};
    int size = sizeof(arr) / sizeof(arr[0]);
    int target = 3;

    int index = lastOccurrenceSearch(
        arr,
        size,
        target
    );

    if (index != -1)
    {
        printf(
            "Last occurrence of %d found at index %d\n",
            target,
            index
        );
    }
    else
    {
        printf(
            "%d not found in array\n",
            target
        );
    }

    return 0;
}

File: Divide-and-Conquer Searching/Binary Search/Last Occurance Search/LastOccurenceSearch.py
def last_occurrence_search(arr, target):
    left = 0
    right = len(arr) - 1
    answer = -1

    while left <= right:
        mid = left + (right - left) // 2

        if arr[mid] == target:
            answer = mid
            left = mid + 1
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return answer


def main():
    arr = [1, 3, 3, 3, 5, 7, 9]
    target = 3

    index = last_occurrence_search(
        arr,
        target
    )

    if index != -1:
        print(
            f"Last occurrence of {target} "
            f"found at index {index}"
        )
    else:
        print(
            f"{target} not found in array"
        )


if __name__ == "__main__":
    main()
